### PR TITLE
611 - Create Non Arrival for Booking

### PR DESCRIPTION
**Pull request by brownbearsoftware**
Implement create Non Arrival endpoint for Bookings + tests

Added own table for the reasons, similar to how Cancellations work

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/factory/NonArrivalReasonEntityFactory.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  fun withIsActive(isActive: Boolean) = apply {
+    this.isActive = { isActive }
+  }
+
+  override fun produce(): NonArrivalReasonEntity = NonArrivalReasonEntity(
+    id = this.id(),
+    name = this.name(),
+    isActive = this.isActive()
+  )
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/factory/NonArrivalReasonEntityFactory.kt:30]: Just thinking about these factories - is there a way we can provide defaults for some of the values? Would save a lot of the repetition in tests where we don't actually care about the values.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt [hunk 2]
```diff
@@ -175,6 +182,45 @@ class PremisesController(
...
+  ): ResponseEntity<Nonarrival> {
+    val premises = premisesService.getPremises(premisesId)
+      ?: throw NotFoundProblem(premisesId, "Premises")
+
+    val booking = bookingService.getBooking(bookingId)
+      ?: throw NotFoundProblem(bookingId, "Booking")
+
+    if (booking.premises.id != premises.id) {
+      throw NotFoundProblem(bookingId, "Booking")
+    }
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:198]: Again, as with #54, it'd be good to reuse this stuff

**Review by pezholio — APPROVED** [review]
LGTM - some similar questions to #54 though

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/factory/NonArrivalReasonEntityFactory.kt (hunk 1 above)
Comment (brownbearsoftware) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/factory/NonArrivalReasonEntityFactory.kt:30]: Yeah so the default values are set when we declare the variables at the top of the class
```
  private var id: Yielded<UUID> = { UUID.randomUUID() }
  private var name: Yielded<String> = { randomStringMultiCaseWithNumbers(6) }
  private var isActive: Yielded<Boolean> = { true }
```

So you don't need to specify any of them in this case if you don't want to, e.g. just do:

```
NonArrivalEntityFactory().produce()
```

which will give you a NonArrival with random id/name and isActive of true

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:198]: Will tackle this alongside the existing endpoints in a follow up PR

**Pull request merged**